src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,215 @@
+#include <emscripten/val.h>
+
+#include <clp/ErrorCode.hpp>
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/ir/LogEvent.hpp>
+#include <clp/ir/LogEventDeserializer.hpp>
+#include <clp/ir/types.hpp>
+#include <clp/streaming_compression/zstd/Decompressor.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <clp/TraceableException.hpp>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <system_error>
+#include <utility>
+
+#include <spdlog/spdlog.h>
+
+#include "ClpJsException.hpp"
+#include "constants.hpp"
+
+using namespace std::literals::string_literals;
+
+// Constants
+namespace {
+constexpr size_t cDefaultNumCharsPerMessage{512};
+constexpr size_t cDefaultNumLogEvents{500'000};
+constexpr size_t cFullRangeEndIdx{0};
+constexpr size_t cLogLevelNone{0};
+}  // namespace
+
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> std::unique_ptr<ClpIrV1Decoder> {
+    auto length{data_array["length"].as<size_t>()};
+    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);
+
+    auto data_buffer{std::make_unique<char const[]>(length)};
+    emscripten::val::module_property("HEAPU8")
+            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.get()));
+
+    auto zstd_decompressor{std::make_shared<clp::streaming_compression::zstd::Decompressor>()};
+    zstd_decompressor->open(data_buffer.get(), length);
+
+    bool is_four_bytes_encoding{true};
+    if (auto const err{
+                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type.."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    std::unique_ptr<ClpIrV1Decoder> ptr(
+            new ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()))
```

Comment:
why not use `make_unique`?

Reply:
The constructor of `ClpIrV1Decoder` is private so it cannot be accessed by `make_unique`.

Anther way is to mark it `friend` but I feel the current way is cleaner.

Reply:
Then I would say we should try to avoid using ptrs if possible (as we discussed in the other comment). Using `new` explicitly is sth we should definitely avoid unless that's the only option (which seems not in JS ffi design)

Reply:
I could be wrong but it might be hard to return a non-pointer. We have the move / copy constructor/assignments deleted which means we can't return by copy.

Reply:
Can we enable move semantics for this class?

Reply:
Hmm... pretty sure I tried yesterday. This is what I got:
```
====================[ Build | ClpIrV1Decoder | Debug ]==========================
/home/junhao/.cache/JetBrains/RemoteDev/dist/0a55911a5a2e1_CLion-2024.1.1/bin/cmake/linux/x64/bin/cmake --build /home/junhao/workspace/clp-ffi-js/cmake-build-debug --target ClpIrV1Decoder -- -j 22
[  5%] Building CXX object CMakeFiles/ClpIrV1Decoder.dir/src/ClpIrV1Decoder.cpp.o
/home/junhao/workspace/clp-ffi-js/src/ClpIrV1Decoder.cpp:81:12: error: call to implicitly-deleted copy constructor of 'ClpIrV1Decoder'
   81 |     return std::move(
      |            ^~~~~~~~~~
   82 |             ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()))
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   83 |     );
      |     ~
/home/junhao/workspace/clp-ffi-js/src/ClpIrV1Decoder.hpp:62:35: note: copy constructor of 'ClpIrV1Decoder' is implicitly deleted because field 'm_data_buffer' has a deleted copy constructor
   62 |     std::unique_ptr<char const[]> m_data_buffer;
      |                                   ^
/home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1/__memory/unique_ptr.h:374:55: note: copy constructor is implicitly deleted because 'unique_ptr<const char[]>' has a user-declared move constructor
  374 |   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 unique_ptr(unique_ptr&& __u) _NOEXCEPT
      |                                                       ^
In file included from /home/junhao/workspace/clp-ffi-js/src/ClpIrV1Decoder.cpp:1:
In file included from /home/junhao/workspace/clp-ffi-js/src/ClpIrV1Decoder.hpp:4:
In file included from /home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:26:
In file included from /home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/val.h:17:
/home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/wire.h:370:20: error: call to implicitly-deleted copy constructor of 'ActualT' (aka 'ClpIrV1Decoder')
  370 |         return new ActualT(std::move(v));
      |                    ^       ~~~~~~~~~~~~
/home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:488:51: note: in instantiation of function template specialization 'emscripten::internal::GenericBindingType<ClpIrV1Decoder>::toWireType<ClpIrV1Decoder>' requested here
  488 |         return internal::BindingType<ReturnType>::toWireType(
      |                                                   ^
/home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:1447:68: note: in instantiation of member function 'emscripten::internal::Invoker<emscripten::internal::rvp::take_ownership, ClpIrV1Decoder, const emscripten::val &>::invoke' requested here
 1447 |         auto invoke = &Invoker<ReturnPolicy, ReturnType, Args...>::invoke;
      |                                                                    ^
/home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:1715:27: note: in instantiation of function template specialization 'emscripten::internal::RegisterClassConstructor<ClpIrV1Decoder (*)(const emscripten::val &)>::invoke<ClpIrV1Decoder, emscripten::return_value_policy::take_ownership>' requested here
 1715 |         invoker::template invoke<ClassType, Policies...>(callable);
      |                           ^
/home/junhao/workspace/clp-ffi-js/src/ClpIrV1Decoder.hpp:72:14: note: in instantiation of function template specialization 'emscripten::class_<ClpIrV1Decoder>::constructor<emscripten::internal::DeduceArgumentsTag, ClpIrV1Decoder (*)(const emscripten::val &), emscripten::return_value_policy::take_ownership>' requested here
   72 |             .constructor(&ClpIrV1Decoder::create, emscripten::return_value_policy::take_ownership())
      |              ^
/home/junhao/workspace/clp-ffi-js/src/ClpIrV1Decoder.hpp:62:35: note: copy constructor of 'ClpIrV1Decoder' is implicitly deleted because field 'm_data_buffer' has a deleted copy constructor
   62 |     std::unique_ptr<char const[]> m_data_buffer;
      |                                   ^
/home/junhao/workspace/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1/__memory/unique_ptr.h:374:55: note: copy constructor is implicitly deleted because 'unique_ptr<const char[]>' has a user-declared move constructor
  374 |   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 unique_ptr(unique_ptr&& __u) _NOEXCEPT
      |                                                       ^
2 errors generated.
make[3]: *** [CMakeFiles/ClpIrV1Decoder.dir/build.make:77: CMakeFiles/ClpIrV1Decoder.dir/src/ClpIrV1Decoder.cpp.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:83: CMakeFiles/ClpIrV1Decoder.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:90: CMakeFiles/ClpIrV1Decoder.dir/rule] Error 2
make: *** [Makefile:124: ClpIrV1Decoder] Error 2
```
from
```
auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> ClpIrV1Decoder {
    auto const length{data_array["length"].as<size_t>()};
    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);

    auto data_buffer{std::make_unique<char const[]>(length)};
    emscripten::val::module_property("HEAPU8")
            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.get()));

    auto const zstd_decompressor{std::make_shared<clp::streaming_compression::zstd::Decompressor>()
    };
    zstd_decompressor->open(data_buffer.get(), length);

    bool is_four_bytes_encoding{true};
    if (auto const err{
                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
        };
        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
    {
        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
        throw ClpJsException(
                clp::ErrorCode::ErrorCode_MetadataCorrupted,
                __FILENAME__,
                __LINE__,
                "Failed to decode encoding type.."
        );
    }

    if (false == is_four_bytes_encoding) {
        throw ClpJsException(
                clp::ErrorCode::ErrorCode_Unsupported,
                __FILENAME__,
                __LINE__,
                "Is not four byte encoding."
        );
    }
    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
            *zstd_decompressor
    )};
    if (result.has_error()) {
        auto const error_code{result.error()};
        SPDLOG_CRITICAL(
                "Failed to decompress: {}:{}",
                error_code.category().name(),
                error_code.message()
        );
        throw ClpJsException(
                clp::ErrorCode::ErrorCode_MetadataCorrupted,
                __FILENAME__,
                __LINE__,
                "Failed to decompress"
        );
    }

    return std::move(
            ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()))
    );
}
```

Reply:
Since the issue is at the unique_ptr, we can change it into shared_ptr. I'm not sure that's a clean approach though.

Reply:
Discussed offline and we eventually found some awkward `clp::ir::LogEventDeserializer` move constructor resolve issue with LLVM toolchains (e.g., `clang++-10`, Emscripten). `g++-10` did find the move constructor but requested an extra source file FileDescriptor.cpp to be compiled because references were found in ReadOnlyMemoryMappedFile.cpp .

We concluded that we would leave the move constructor to be default while delete the move assignment operator to get rid of the compile warnings. The behaviour is also documented inline in the move constructor / assignment definitions.